Add cmake build scripts and appveyor.yml

alex85k opened the pull request:
Here is my attemp to make cmake build scripts available for main repository.
I have modified .travis.yml to build one version with cmake too. 

To make appveyor work like here https://ci.appveyor.com/project/alex85k/osm2pgsql , you need to login and enable it for main repository.

Notes:
 - Have two warnings in cmake build, can not apply -Werror (some flags??)
https://travis-ci.org/openstreetmap/osm2pgsql/jobs/85616194
 - In appveyor.yml there is ACCOUNT variable and `curl -T` - it is uploading windows build to my shared folder https://yadi.sk/d/ACuCgfNIRUULH and should be better replaced in production
 - In appveyor.yml there is a download of Windows dependenices from my Dropbox. Better to be placed somewhere else :)

pnorman commented on CMakeLists.txt:
```diff
@@ -0,0 +1,180 @@
+endif()
+
+# first try to find the version
+find_package(Boost 1.48 REQUIRED COMPONENTS system chrono filesystem thread date_time)
```
1.50 is the minimum boost version

pnorman replied:
configure.ac only requires system, filesystem, and thread. are chrono and date_time new?

alex85k replied:
I guess it was introduced before C++ migration. Will remove them now.

alex85k replied:
On windows with static boost I could not get rid of unneeded libraries, will add them for MSVC only.

pnorman commented:
Do the readme instructions need adjusting?

pnorman commented:
Don't I need to enable something with appveyor before appveyor.yml makes sense?

pnorman commented on .travis.yml:
```diff
@@ -28,6 +28,9 @@ matrix:
       env: BUILDCXXFLAGS='-Werror -pedantic -Wall -std=c++11'
+    - os: linux
+      compiler: gcc
+      env: BUILDCXXFLAGS='-pedantic -Wall -std=c++11' BUILDSYSTEM='cmake'
```
this will cause two linux instances running gcc to be started

alex85k replied:
One of them uses old autotools...

pnorman commented:
You need to `git rm` autogen.sh, m4/, configure.ac, and any other files for the autotools build system

pnorman commented on CMakeLists.txt:
```diff
@@ -0,0 +1,180 @@
+
+
+#############################################################
+# Build there library and executale file
```
s/executale/executable

alex85k replied:
OK :)

pnorman commented on CMakeLists.txt:
```diff
@@ -0,0 +1,180 @@
+endif()
+
+# Just in case user installed RPMs from http://yum.postgresql.org/
+list(APPEND PostgreSQL_ADDITIONAL_SEARCH_PATHS /usr/pgsql-9.3 /usr/pgsql-9.4)
```
Should this go upstream into the cmake script for postgres?

alex85k replied:
Not sure. By default even Ubuntu can not find installed PostgreSQL. Is it better to include all pathes in cmakelists.txt to avoid "not compile" tickets?

alex85k replied:
Let us drop the paths for now and describe it in readme for different systems...

pnorman replied:
We can't be requiring options on common standard systems. If the standard cmake scripts don't pick up postgres on debian or ubuntu, that's a bug with them which we'll need to work around, probably in CMakeLists

alex85k replied:
Agree. It should be fixed in newest cmake but until now we'll need some workarounds.

alex85k replied:
Updated the branch.

pnorman commented on CMakeLists.txt:
```diff
@@ -0,0 +1,180 @@
+  message(FATAL_ERROR "In-source builds are not allowed, please use a separate build directory like `mkdir biuld && cd build && cmake ..`")
+endif()
+
+execute_process (COMMAND git describe OUTPUT_VARIABLE PACKAGE_VERSION OUTPUT_STRIP_TRAILING_WHITESPACE)
```
What's the git describe for? It should build with just the download, not a git checkout

alex85k replied:
This is version determining. Should we just hardcode it in CMakeLists.txt like was ``AC_INIT([osm2pgsql],[0.89.0-dev])`` or do something more intelligent?

pnorman commented on CMakeLists.txt:
```diff
@@ -0,0 +1,180 @@
+OPTION(BUILD_TESTS "Build with tests" OFF)
+
+if("${PROJECT_SOURCE_DIR}" STREQUAL "${PROJECT_BINARY_DIR}")
+  message(FATAL_ERROR "In-source builds are not allowed, please use a separate build directory like `mkdir biuld && cd build && cmake ..`")
```
s/biuld/build/

alex85k replied:
Yes)

pnorman commented:
> Have two warnings in cmake build, can not apply -Werror (some flags??) https://travis-ci.org/openstreetmap/osm2pgsql/jobs/85616194

libosmium should be included with `-isystem`

-----

CMakeLists.txt doesn't have a consistent style for if statements, it uses `if(foo)`, `if (foo)`, `if ( foo )`. Unless there's a specific cmake convention otherwise, `if (foo)` is consistent with the C++ code

pnorman commented:
How do you run a single `make check` test? With autotools it's `make check TESTS=test/test-output-pgsql-schema`.

libosm2pgsql.a shouldn't be installed with `make install`, it exists only for the tests. default.style should be installed. nodecachereader is currently installed, but I'd be okay dropping that from the install if there are solid reasons. @woodpeck found it had been broken for over a year with 64-bit IDs, so clearly no one is using it.

alex85k commented:
Single make check tests is done by 
```
ctest -VV -R test-output-pgsql-schema
or
ctest -VV -R test-output-*
```
(-VV is verbose)
Can try to put this into custom check target.
----


Thank you for review! Will re-push with updates. 
I think it is too early to remove autotools without public testing cmake for some time. Maybe we can keep them both for some period?

alex85k commented:
Next version is ready. Git is not needed for building, fixed code style, install man and styles, build with -Werror by cmake with system Osmium libs, min boost=1.50 with date_time and chrono on Windows only. Did not remove autotools and edit readme for now..

pnorman commented:
> I think it is too early to remove autotools without public testing cmake for some time. Maybe we can keep them both for some period?

Two build systems is likely to cause trouble. Instead, we should test the cmake stuff in the branch, and toss autotools.

alex85k commented:
No problem... But many tickets are likely to appear :)

alex85k commented:
Here it is. PostgreSQL search is fixed by fake PostgreSQL_TYPE_INCLUDE_DIR variable - it is not needed :)
Additional prexes for RPMs from http://yum.postgresql.org/ inserted.

By the way, most Linux compatibility changes in my script were initiated by @DerDakon in his PR. Maybe he could help with support here too.

DerDakon commented on .gitignore:
```diff
@@ -64,3 +64,4 @@ libosm2pgsql.la
 
 docs/html
 docs/latex
+build
```
Missing newline

alex85k replied:
ok)

DerDakon commented:
What is the actual problem with PostgreSQL? Is it required or not? If not, then the REQUIRED should be removed from the CMakeLists.txt, then there is no need to fake anything. This would for sure cause trouble for anyone that has PostgreSQL in a non-standard location and needs that. And it is also possible with CMake to pass additional flags like -Werror from the outside, just call

 CXXFLAGS=-Werror cmake ...

alex85k commented:
Postgresql client is required, but pg_types.h is not needed. Standard check fails if it can not find it (no such file on travis without additional packages). On other system PostgreSQL_TYPE_INCLUDE_DIR is also not needed, so it should be no problem defining it.

alex85k commented:
@pnorman : would you prefer optional CXXFLAGS like ```-Werror -pedantic -Wall`` as variable or inside CMakeLists.txt?

DerDakon commented:
It doesn't make much sense to hardcode compiler specific flags in a CMakeLists.txt as it can easily break the build if a different compiler is used, so I would vote to put in the test configuration files.

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,200 @@
+# Just in case user installed RPMs from http://yum.postgresql.org/
+list(APPEND PostgreSQL_ADDITIONAL_SEARCH_PATHS /usr/pgsql-9.3 /usr/pgsql-9.4)
+
+if ("${PROJECT_SOURCE_DIR}" STREQUAL "${PROJECT_BINARY_DIR}")
```
Remove the dereference of the variables, this can cause trouble if the content inside any of these variables is the name of another variable. Just use

 if (PROJECT_SOURCE_DIR STREQUAL PROJECT_BINARY_DIR)

alex85k replied:
Thank you. First time I hear about such tricks... (copy-pasted in multiple projects)

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,200 @@
+message(STATUS "Building osm2pgsql ${PACKAGE_VERSION}")
+
+if (NOT CMAKE_BUILD_TYPE)
+  set(CMAKE_BUILD_TYPE=Release)
```
This sets a variable CMAKE_BUILD_TYPE=Release, you want to replace the = with a space to set CMAKE_BUILD_TYPE.

alex85k replied:
Yes, long-lasted bug.

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,200 @@
+  set (CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} /STACK:30000000")
+else()
+  add_definitions(-Wall)
+  add_definitions(-std=c++11)
```
Depending on the minimum CMake version that will be required this can be set by CMake (when at least 3.1 is present).

alex85k replied:
Travis CI and many distros have 2.8, I had to downgrade intentionally.

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,200 @@
+#############################################################
+
+if (NOT EXTERNAL_LIBOSMIUM)
+  set(OSMIUM_INCLUDE_DIR "contrib/libosmium")
```
This should become "${CMAKE_CURRENT_SOURCE_DIR}/contrib/libosmium" so it is an absolute path.

alex85k replied:
Agree.

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,200 @@
+  endif()
+endif()
+
+if (WIN32)
```
Simply move this below the set(LIBS) below and just append to the LIBS variable.

alex85k replied:
And I'll remove ``if (NOT PostgreSQL_FOUND)`` inserting help for getting wingetopt.

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,200 @@
+   file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/unistd.h "// empty header\n")
+endif()
+
+include_directories(${CMAKE_CURRENT_BINARY_DIR}) # for generated files
```
set(CMAKE_INCLUDE_CURRENT_DIR On)

This will automatically add the current source and binary dir to the include path.

alex85k replied:
Nice...

alex85k replied:
Happens that breaks test compilation (dirs changes). So leaving as is.

DerDakon replied:
Ah, yes, those includes are not propagated into subdirectories.

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,200 @@
+endif()
+
+include_directories(${CMAKE_CURRENT_BINARY_DIR}) # for generated files
+include_directories(contrib/libosmium)
```
Should be removed, the libosmium include path was already set above, this would only duplicate it and might cause troube when an external libosmium is used.

alex85k replied:
Yes, that's just a forgotten duplicate. Removing...

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,200 @@
+include_directories(contrib/libosmium)
+include_directories(${CMAKE_CURRENT_SOURCE_DIR})
+
+file(GLOB osm2pgsql_lib_SOURCES RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} *.cpp *.hpp)
```
This is bad practice, as new cpp or hpp files will not trigger a project rebuild because CMake can't detect them automatically. Better list them explicitely.

alex85k replied:
Let it be. I dislike big lists but maybe it will help someone avoid weird errors...

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,200 @@
+endif()
+
+# Emulate make check, use like "TESTS=test-middle make check"
+add_custom_target(check COMMAND ${CMAKE_CTEST_COMMAND} -VV -R ${TESTS})
```
CMake automatically adds "make test".

alex85k replied:
Do other developers need separate verbose variant?

DerDakon commented on README.md:
````diff
@@ -48,7 +48,7 @@ Both GCC 4.8 and Clang 3.4 meet this requirement.
 
 ```sh
-sudo apt-get install autoconf automake libtool make g++ pkg-config libboost-dev \
+sudo apt-get install make cmake g++ pkg-config libboost-dev \
````
Is pkg-config still needed? For Fedora below it is removed.

alex85k replied:
Need testing, that strings of readme are approximate for now.

alex85k replied:
I remembered that pkg-config is used by cmake sometimes.

DerDakon replied:
The Find* CMake modules can use pkg-config to figure out where the stuff is. None of the modules used here does this (yet). So either way is fine, with or without pkg-config, but it should be consistent between the different platforms IMHO.

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,247 @@
+  set(OSMIUM_INCLUDE_DIR "${CMAKE_CURRENT_SOURCE_DIR}/contrib/libosmium")
+endif()
+
+include_directories(${CMAKE_SOURCE_DIR})
```
That will cause problems if anyone else chooses to have osm2pgsql build as part of a larger CMake project, then this directory may end up being not the top level directory. Use

include_directories(${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_BINARY_DIR})

alex85k replied:
:ok:

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,247 @@
+message("Active compiler flags:" ${CMAKE_CXX_FLAGS})
+
+#############################################################
+# Build there library and executable file
```
"Build the ..."?

alex85k replied:
:abcd: :ok:

DerDakon commented on README.md:
```diff
@@ -65,25 +65,30 @@ sudo yum install gcc-c++ automake libtool pkgconfig boost-devel \
-    LUA_LIB=`pkg-config --libs lua-5.2` ./configure && gmake && gmake install
+    make && make install
+
+Please see `./configure --help` for more options on how to control the build
```
configure is gone.

alex85k replied:
Almost gone :smile:

DerDakon commented on cmake/config.h.in:
```diff
@@ -0,0 +1,43 @@
+#define snprintf _snprintf
+#endif
+
+typedef __int64 ssize_t;
```
This is wrong for 32 bit Windows. No idea if anyone still cares. But if you do, there seems to be a correct define on Windows which you can use: https://msdn.microsoft.com/de-de/library/windows/desktop/aa383751(v=vs.85).aspx

alex85k replied:
We are lucky that 32 bit is certainly unsuppored, therei a specific check in CMakeLists.txt.

pnorman replied:
I'm skeptical if 32-bit windows works with osm2pgsql. It'd be nice to have it work, but just getting Windows working properly is enough of a goal for now

DerDakon replied:
The fix is simple, so I would just include it:

```C
#include <BaseTsd.h>
typedef SSIZE_T ssize_t;
```

alex85k replied:
Forgot about this, will add to PR tomorrow (maybe with building 32 bit on my pc)

DerDakon commented on tests/CMakeLists.txt:
```diff
@@ -0,0 +1,49 @@
+  add_executable(${test_name} ${test})
+  target_link_libraries(${test_name} osm2pgsql_lib common-pg middle-tests)
+  add_test(NAME ${test_name} COMMAND ${test_name}
+    WORKING_DIRECTORY ${CMAKE_SOURCE_DIR})
```
Better use ${osm2pgsql_SOURCE_DIR} here to be sure to refer to the correct directory. This is automatically set to the directory where project(osm2pgsql) is called.

alex85k replied:
One more useful fact about CMake, thank you!
Fixing...

DerDakon commented on tests/CMakeLists.txt:
```diff
@@ -0,0 +1,49 @@
+find_package(PythonInterp)
+
+if (PYTHONINTERP_FOUND)
+  add_test(NAME regression-test-pbf COMMAND python tests/regression-test.py -f tests/liechtenstein-2013-08-03.osm.pbf
```
If you use FindPythonInterp, you should also use ${PYTHON_EXECUTABLE} instead of "python".

alex85k replied:
Hm, I though I did it.  Will fix.

alex85k replied:
Yes, we'll use better Python :snake:   . And ${osm2pgsql_SOURCE_DIR} in nex tine too.

alex85k commented:
Almost did it.
We can also run tests on Travis if we isolate Flat nodes related ones :)

alex85k commented:
> What do you need that for? Noone uses Dart anymore AFAIKT. In doubt, use CDash. Otherwise just drop this line and the next one.

I needed timeout, becuse debug tests on Windows are very slow.

alex85k commented on tests/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+  target_link_libraries(${test_name} osm2pgsql_lib common-pg middle-tests)
+  add_test(NAME ${test_name} COMMAND ${test_name}
+    WORKING_DIRECTORY ${osm2pgsql_SOURCE_DIR})
+  set_tests_properties(${test_name} PROPERTIES TIMEOUT ${TESTING_TIMEOUT} ENVIRONMENT srcdir=${osm2pgsql_SOURCE_DIR})
```
Will this set timeout correctly instead of Dart subsystem?

DerDakon replied:
Yes, I think so.

pnorman commented:
> @pnorman : would you prefer optional CXXFLAGS like -Werror -pedantic -Wall as variable or inside CMakeLists.txt?

Outside - ideally we shouldn't be hard-coding *any* flags but telling cmake we want all warnings. We shouldn't be specifying `-Werror` in the build system.

Within Travis is different - we want to check code quality pre-merge, so any warnings should cause the Travis build to fail and we want the compiler to be very picky hence `-Werror -Wall -pedantic` in order to stop problems from sneaking in.

pnorman commented on README.md:
````diff
@@ -65,25 +65,29 @@ sudo yum install gcc-c++ automake libtool pkgconfig boost-devel \
 
 ```sh
-pkg install devel/git devel/autoconf devel/automake devel/gmake devel/libtool \
+pkg install devel/git devel/cmake
````
The lines below need rewrapping

alex85k replied:
The package list needs checking too. Will try on clean 10.2 release.

alex85k replied:
Will do it. The list is for building not for running tests, I guess?

pnorman replied:
Yes, only build and runtime dependencies should go here, not dev dependencies. That being said, we do need to document the development dependencies somewhere (#182)

pnorman replied:
I was thinking we still needed gmake, but I guess the Makefiles now work with BSD make.

I'm pretty sure the list includes everything required and nothing extra, with the possible exception of pkgconf which has been raised elsewhere

pnorman commented on README.md:
````diff
@@ -65,25 +65,29 @@ sudo yum install gcc-c++ automake libtool pkgconfig boost-devel \
 ```
 
-Then you should be able to bootstrap the build system:
+CMake 2.8.7 is needed to build oslm2pgsql. First generate the makefiles in separate folder:
````
s/oslm2pgsql/osm2pgsql/. Also, don't I wouldn't mention the cmake version

pnorman commented on README.md:
```diff
@@ -65,25 +65,29 @@ sudo yum install gcc-c++ automake libtool pkgconfig boost-devel \
 
-    ./autogen.sh
+    mkdir build && cd build
+    cmake .. -G "Unix Makefiles"
```
is `-G "Unix Makefiles"` required

alex85k replied:
On some systems it does... But maybe it was on Windows :)

Will try to remove it from travis.

DerDakon replied:
CMake will use the default generator if none is specified. This is always Makefiles on Unix systems, but is usually one MSVC variant on Windows systems.

pnorman replied:
Is there a need to change from the default on Windows?

alex85k replied:
Usually it is needed (many Visual Studios, 32/64 bits are also there, someone prefer nmake makefiles), but this does not affects Travis CI and readme. apvveyor.yml have such option for Visual Studio 64 bit build.

On Windows you have to specify many options dependint on machine beacuse there are no standard pakages and locations for Boost, Expat, etc. I keep them in https://github.com/alex85k/winbuilds , there are other specialized coolections.

pnorman commented on README.md:
```diff
@@ -65,25 +65,29 @@ sudo yum install gcc-c++ automake libtool pkgconfig boost-devel \
 
-Please see `./configure --help` for more options on how to control the build
-process.
+If some packages are not found, specify them directly. For example, on old Ubuntu you may need to add the following option:
```
I'd rather not specify excess custom install information for EOL distros, and we need to work without any additional options on Ubuntu 14.04

alex85k replied:
Currenty it should work without options. Have no 14.04 to check, however.

pnorman commented on README.md:
```diff
@@ -65,25 +65,29 @@ sudo yum install gcc-c++ automake libtool pkgconfig boost-devel \
+    cmake .. -G "Unix Makefiles" -DPostgreSQL_TYPE_INCLUDE_DIR=/usr/include/postgresql/9.4/server/catalog
 
-    LUA_LIB=`pkg-config --libs lua-5.2` ./configure && gmake && gmake install
+Wnen the makefiles are generated successfully, run the standard GNU build install:
```
s/Wnen/When/, but the wording here seems awkward and reworking it might be better

alex85k commented:
FreeBSD 10.2 build from readme.md was successful.

alex85k commented:
Renamed TEST_BUILD option to classical BUILD_TESTS .

alex85k commented:
I have introduced test label and allowed 6 simple ones run on Travis+Appveyor.
```
  ctest -VV -L NoDB
```
(NoDB is a label for non-db-connecting-tests).
3 questions:
  - Was srcdir environment variable I deleted from xml2 test really needed?
  - Does test-output-pgsql require 20 Gb flat-nodes file or not?
  - Why 3 simple tests get segfault on CLang?  :)

alex85k commented:
By the way, 30Gb filesystem is not enough to run   test-middle-flat and test-output-pgsql :)
Other tests seem to pass...

pnorman commented:
> By the way, 30Gb filesystem is not enough to run test-middle-flat and test-output-pgsql. Is it normal?
> Other tests pass in reasonable time.

Do you have the changes introduced in #458 in this branch?

-----

> Why can 3 simple tests get segfault on Clang? I have only enabled them, not changed the code ... :)

Good question. I hit this before, and I can't reproduce it off of Travis. It works with GCC on Travis, and Clang on Ubuntu 12.04 and 14.04. I recommend running tests single-threaded, and if that doesn't work, disabling those tests for now.

alex85k commented:
> Do you have the changes introduced in #458 in this branch?
Yes. Actually, I link common-cleanup.cpp to all of the tests to be sure...

alex85k commented:
Considerin failing tests: I had similar problems on Appveyor when some of the dlls were not in the path. Exactly with test-options-parse . Could the complex gcc-built library conflict with clang-based executable? (on Windows it usually does)

alex85k commented:
Seems the bug on Clang is triggered when boost_tread library is loaded
https://travis-ci.org/alex85k/osm2pgsql/jobs/86131797
Some clang-incompatible package problem?

pnorman commented:
> Seems the bug on Clang is triggered when boost_thread library is loaded
> https://travis-ci.org/alex85k/osm2pgsql/jobs/86131797

Where do you see that?

If that is the case, we might be missing a flag.

> Some clang-incompatible package problem?

It builds with those packages and clang, just not on Travis

alex85k commented:
Here:
https://travis-ci.org/alex85k/osm2pgsql/jobs/86131797

Not our code in stack but libdl etc:
```
76: ==7239== Command: /home/travis/build/alex85k/osm2pgsql/build/tests/test-parse-xml2
76: ==7239== 
76: ==7239== 
76: ==7239== Process terminating with default action of signal 11 (SIGSEGV)
76: ==7239==  General Protection Fault
76: ==7239==    at 0x525C947: boost::exception_ptr boost::exception_detail::get_static_exception_object<boost::exception_detail::bad_exception_>() (in /usr/lib/x86_64-linux-gnu/libboost_thread.so.1.55.0)
76: ==7239==    by 0x5257540: ??? (in /usr/lib/x86_64-linux-gnu/libboost_thread.so.1.55.0)
76: ==7239==    by 0x400F305: call_init.part.0 (dl-init.c:85)
76: ==7239==    by 0x400F3DE: _dl_init (dl-init.c:52)
76: ==7239==    by 0x40016E9: ??? (in /lib/x86_64-linux-gnu/ld-2.15.so)
```

alex85k commented:
Added final TESTS_VALGRIND option to enable memory checking in tests if needed.
Temporary enabled it for failing clang. It seems that boost runtime on Travis  is not quite clang-compatible. Most likely will just disable the tests for that platform in travis.yml.

pnorman commented:
> Added final TESTS_VALGRIND option to enable memory checking in tests if needed

I'd keep this out for now, let's keep the cmake PR from growing too much in scope. We're not building the failing tests on travis right now, and I think it's okay to keep it that way.

pnorman commented on cmake/FindLua52.cmake:
```diff
@@ -0,0 +1,82 @@
+# See the License for more information.
+#=============================================================================
+# (To distribute this file outside of CMake, substitute the full
+#  License text for the above reference.)
```
We need to substitute the full license

DerDakon replied:
And I still strongly suggest to use the module version that was actually accepted upstream.

alex85k replied:
If we want it to work on Travis, we should support 2.8.7 Cmake. Does it have working module?

DerDakon replied:
I mean to copy the upstream FindLua.cmake module here instead of the FindLua52.cmake.

alex85k replied:
Many options are not supported on old cmake, I tried with Postgre. WIll check, however...

alex85k replied:
@DerDakon : New FindLua.cmake is your module added to 3.0.0.
It may not work on 2.8.7 but I'll try.

alex85k replied:
```
CMake Error at cmake/FindLua.cmake:164 (include):
  include could not find load file:

    D:/winbuilds_d/osm2pgsql/cmake/FindPackageHandleStandardArgs.cmake
```
On new cmake... How to tell it to use system FindPackageHandleStandardArgs?

DerDakon replied:
include(FindPackageHandleStandardArgs)

alex85k replied:
``include(${CMAKE_CURRENT_LIST_DIR}/FindPackageHandleStandardArgs.cmake)``
will still break.. Had to comment you. Does author agree? :)

DerDakon replied:
replace that with "include(FindPackgeHandleStandardArgs)". This will pick the module from the default search path (i.e. no path, no extension).

alex85k replied:
Done. :+1:

pnorman commented on tests/CMakeLists.txt:
```diff
@@ -0,0 +1,77 @@
+set(TESTING_TIMEOUT 2000)
```
What units is this in?

DerDakon replied:
Seconds.

pnorman replied:
let's use a time like 1200, which is 20 minutes, and long enough for the python regression test on the slowest VM I've used

alex85k replied:
It was not debug version on Windows, I guess :)
Flat-nodes test can take even more at 20-30 mb/s 20Gb generating several times. I suggest to make it a parameter, 1200 by default (to be sure Travis will not hang).

alex85k replied:
Added, if needed  we'll use
```
cmake .. -DBUILD_TESTS -DTESTING_TIMEOUT=3600
```

pnorman commented on tests/test-parse-xml2.cpp:
```diff
@@ -128,14 +128,8 @@ void assert_equal(uint64_t actual, uint64_t expected) {
-  }
-
-  std::string inputfile = std::string(srcdir) + std::string("/tests/test_multipolygon.osm");
+  std::string inputfile = "tests/test_multipolygon.osm";
```
Where is srcdir defined?

alex85k replied:
It was in environment, cmakelists.txt in our case. All other tests did not use it.

DerDakon commented:
For future experiments: the valgrind stuff can easily be achieved with just CMake itself: add "include(CTest)" in line 238 of CMakeLists.txt and then "make ExperimentalMemCheck" after "make test".

pnorman commented:
.gitignore needs updating to remove the autotools files. Aside from that, the travis stuff, and other comments, it looks okay, but I want to put out a call for wider testing on a different variety of distributions.

alex85k commented:
Will include that magical CTest, remove valgrind and clean ignore. 

By the way, we completely ignore geos-fallback now. It it still needed?

alex85k commented:
Commented
```
include(${CMAKE_CURRENT_LIST_DIR}/FindPackageHandleStandardArgs.cmake)
```
in upstream FindLua.cmake . Is it OK?

Also pre-created build folder with .gitignore and build.sh to simplify build. Is it needed?

alex85k commented:
Also disabled test on Linux-clang . That's all for now...

alex85k commented:
Diffrerences from verbose make (skipping identical parts):
```
CMake (Release build, default):
g++  -DBOOST_TEST_DYN_LINK  -D_FILE_OFFSET_BITS=64  -D_LARGEFILE_SOURCE
  -O3 -DNDEBUG .....

CMake (Debug build):
g++  -DBOOST_TEST_DYN_LINK  -D_FILE_OFFSET_BITS=64  -D_LARGEFILE_SOURCE
  -g .....

Autotools:
g++ -isystem geos-fallback
 -g -O2 -Wall ......

```
Are we missing -Wall  ? Does anyaone need geos-fallback or -g by defaut?

pnorman commented:
> Are we missing -Wall

We should be specifying it for all builds, including non-travis. With autotools we're using `AX_CFLAGS_WARN_ALL` and `AX_CXXFLAGS_WARN_ALL`

`-D_LARGEFILE_SOURCE`

node-presistent-cache.cpp has `#define _LARGEFILE64_SOURCE     /* See feature_test_macrors(7) */`. I don't know what it does, so I'm not sure if it's best to globally define it.

DerDakon commented:
See http://www.delorie.com/gnu/docs/glibc/libc_13.html for _LARGEFILE64_SOURCE: it enables API to work with files >2GB on 32 bit systems.

alex85k commented:
Let's add ``-Wall`` to cmake then and put ``#define _LARGEFILE_SOURCE`` in ``config.h``.
What about optimizations, debugging info and geos-fallback?

pnorman commented:
> put #define `_LARGEFILE_SOURCE` in `config.h`.

I think we could just leave it where it is, it's only needed for `lseek64` which is only in the flat-nodes code, and it only matters on 32-bit machines.

> Let's add `-Wall` to cmake then

Is there not a compiler-independent way of enabling warnings? Not all compilers use `-Wall` for all warnings.

> What about optimizations, debugging info

Our autotools setup doesn't really have release vs debug builds, so the cmake behavior seems sane. We might need to come back and tweak it. The one thing I'm not sure about is omitting debug info for release builds.

We should document how to do both a debug and a release build.

> geos-fallback?

Not needed on Ubuntu 12.04 anymore, so I'm fine with completely removing the files and any reference to them, but it'll need testing. If you don't want to do that in the scope of this PR, you could just replicate the current behavior.

alex85k commented:
There are following changes in last push:
   - updated readme.md for RPM-based systems and build options
   - switched to RelWithDebugInfo configuration by default (optimizations + debug info)
   - added ``-Wall`` for all non-MSVC compilers. GCC, Intel and clang knows it, is there anything else C++11?
   - removed geos-fallback . No code changes are needed and using it as -system can potentially cause problems. Let us test all compilation problems here without asking people to check the build twice :)

alex85k commented:
Also
  - updated build/build.sh to build tests by default. By the way, is OK to keep that script and pre-created folder? Maybe add test.sh there too or remove?
  - added proj-epsg rpm to readme (was failing options test on CentOS)

pnorman commented:
> updated build/build.sh to build tests by default. By the way, is OK to keep that script and pre-created folder? Maybe add test.sh there too or remove?

I'm not sure why we need it

pnorman commented:
Are we using the cmake lua script from upstream?

alex85k commented:
> I'm not sure why we need it

I am not sure too, Maybe for people unfamiliar with cmake? I just used it to fasten the build process on VMs (no need to do mkdir etc.). Different options can be placed in comments there too. But it can be removed.

>  Are we using the cmake lua script from upstream?
Yes, the latest variant from git: https://github.com/Kitware/CMake/blob/master/Modules/FindLua.cmake
. One line is patched to work outside system modules dir, @DerDakon is the script real author :)

pnorman commented:
> I am not sure too, Maybe for people unfamiliar with cmake? I just used it to fasten the build process on VMs (no need to do mkdir etc.). Different options can be placed in comments there too. But it can be removed.

I'd get rid of it.

-----

I'm going to put out a call for help testing this PR on different systems.

alex85k commented:
Removed .sh, fixed travis, appveyor and readme to create build folder.

alex85k commented:
If anyone asks, here are the scripts to build and bits of configuration:
https://yadi.sk/d/64G6cNtzjuWrX

pnorman commented on cmake/config.h.in:
```diff
@@ -0,0 +1,43 @@
+#ifdef _MSC_VER
+#if _MSC_VER < 1900
+#define snprintf _snprintf
+#endif
```
This might not be required anymore. Will versions <1900 compile C++11?

alex85k replied:
Agree. Will give it a try with other VS cleanups, trying 32bit too.

The only released msvc that supports c++11 is from VS2015. Partial support is from vs2013 november ctp, but I doubt we should support it.

alex85k commented:
Get rid of _snprintf completely (certainly need VS2015 to build c++11), used SSIZE_T from BaseTsd.h as @DerDakon adviced. Now trying to make 32bit windows build.

alex85k commented:
32bit Windows not only builds, but also passes the tests! (except out-of-hard-drive flat-nodes for now)
Will add it to Appveyor.yml configuration. 

By the way, there is even a cahnce we can run some tests on Appveyor automatically!
http://help.appveyor.com/discussions/suggestions/409-postgis-extension-for-postgresql

alex85k commented:
Rebased to latest. The tests on Appveyor now pass, so the pre-merge setup of appveyor in the main repository can already be done (tick 'ignore branches without appveor.yml' in settings).

pnorman commented:
AppVeyor enabled. I'm not going to worry too much about it before merging.

alex85k commented:
Thank you. For now fixing appveyor errors an warnings is on me :)

gdt commented:
Works on on NetBSD 7 amd64 with gcc 4.8.4, with no warnings.
When building on NetBSD 6 i386, which has gcc 4.5 native, but configured with gcc 4.8.4 in $PATH, I get complaints 
/home/gdt/SOFTWARE/GEO/osm2pgsql/contrib/libosmium/osmium/io/bzip2_compression.hpp:91:30: error: ‘to_string’ is not a member of ‘std’
                     error += std::to_string(bzlib_error);
                              ^
and this might be a cmake issue; it's unclear to me how compiler selection and the right includes/libs get set.   Has this been tested on other systems where the system compiler is < 4.8, but 4.8 is available via a package at a different path?

pnorman commented:
> Has this been tested on other systems where the system compiler is < 4.8, but 4.8 is available via a package at a different path?

Travis runs on Ubuntu 12.04 and has to use a GCC version from PPAs.

Could you try checking out master and building with autotools with `gmake V=1` and seeing what command line is used for building? If GNU Make is just `make` on NetBSD, use that instead.

alex85k commented:
Had simlar problems compiling for my non-root RHEL 4 system. 
Compilers can be changed by setting CC=path/to/gcc CXX=path/to/g++ or by specifying options
```
cmake -DCMAKE_C_COMPILER=... -DCMAKE_CXX_COMPILER=...
```
The path is almost ignored!

alex85k commented:
@gdt You may also need to set CMAKE_PREFIX_PATH to your software in home folder, elsewhere cmake can prefer system libraries.

DerDakon commented:
The first question about the NetBSD problem is: which compiler did CMake find? It will print it at the first lines of it's output, giving the version number. Every compiler should know about it's internal include dir, so I suspect it is something like the wrong (global?) STL is used. Do you have multiple libstdc++.so libraries on your system?

alex85k commented:
My report from exotic system: RedHat 6.1, no root, with installed old gcc, old boost , old lua etc.
Done this (with multiple trial-and-errors):
 - Built fresh cmake 3.3.1
 - Built gcc 5.2.0 in custom directory
 - Built all the libraries with it, placing libs and includes together with gcc. (bzip2, zlib, expat, proj, geos 3.5.0 (no SVN!))
 - Built Boost 1.59 and Postgresql 9.4.5 in separate folders 
 
To get rid of system libs for cmake had to do this:
```
GCC_HOME=/gpfs/soft/gcc/5.2.0-rh6
export LD_LIBRARY_PATH=${GCC_HOME}/lib64:${GCC_HOME}/lib:$LD_LIBRARY_PATH
export LD_RUN_PATH=${GCC_HOME}/lib64:${GCC_HOME}/lib:$LD_LIBRARY_PATH
export PATH=${GCC_HOME}/bin:$PATH
export C_INCLUDE_PATH=${GCC_HOME}/include
export CPLUS_INCLUDE_PATH=${GCC_HOME}/include

export CC=/gpfs/soft/gcc/5.2.0-rh6/bin/gcc
export CXX=/gpfs/soft/gcc/5.2.0-rh6/bin/g++
export CMAKE_PREFIX_PATH=/gpfs/soft/postgresql/v945_gcc520
cmake .. -DBOOST_ROOT=/gpfs/soft/boost/gcc520  -DCMAKE_BUILD_TYPE=Release -DBoost_ADDITIONAL_VERSIONS="1.57;1.58;1.59" -DBUILD_TESTS=ON
make -j8
ctest -VV -L NoDB
```
It worked!

Notes:
  - gcc from git 3.5.0 branch is incompatible - GeometryFactory had been made protected
  - cmake does not find boost 1.59 automatically for now - we may need to add it to CMakeLists.txt
  - all the custom-built lib should be placed together with custom gcc or listed in export CMAKE_PREFIX_PATH .

alex85k commented:
One more thing:
  - with my self-made Boost linking to ``rt`` system library was needed. Cmake did not determine it.
```
thread.cpp:(.text+0x5e5): undefined reference to `clock_gettime'
``` 
Lua needed it too and also -ldl... Had to add linker flags manually
```
  -DCMAKE_EXE_LINKER_FLAGS="-lrt -ldl" 
```
I guess this should be fixed in cmake, but do not know how.

DerDakon commented:
Something like that

```CMake
target_link_libraries(foo ${CMAKE_DL_LIBS})
if (CMAKE_SYSTEM_NAME STREQUAL Linux)
    target_link_libraries(foo rt)
endif ()
```

DerDakon commented:
Which part of boost needs rt? In that case the upstream FindBoost.cmake should probably be fixed to include that. I wonder why neither boost nor Lua linked to that themself.

alex85k commented:
My custom-built boost thread used rt because I specified ``runtime-link=dynamic`` when building boost. Not sure if it can be detected.

Am I right that linking to ${CMAKE_DL_LIBS} and rt is safe on Linux, even if it is actually not needed?

alex85k commented:
I also found that some of the boost libraries in CMakeCache.txt are still from /usr. That may cause a conflict and I doubt the CMake is to blame here :) (I better build full debug+release boost)

DerDakon commented:
Using CMAKE_DL_LIBS is safe on all platforms, that library is simply empty e.g. on Windows. Linking to rt on platforms where it is present but when it is not needed by the program will not cause any errors, in most cases it would even be dropped by the linker if it is not necessary.

alex85k commented:
Thanks for help! Committed the workarounds and readme updates. With them cmake line is shorter:
```sh
export CC=/gpfs/soft/gcc/5.2.0-rh6/bin/gcc CXX=/gpfs/soft/gcc/5.2.0-rh6/bin/g++
export CMAKE_PREFIX_PATH=/gpfs/soft/postgresql/v945_gcc520
cmake -DBOOST_ROOT=/gpfs/soft/boost/gcc520 -DLUA_LIBRARY=/gpfs/soft/gcc/5.2.0-rh6/lib/liblua.a -DBUILD_TESTS=ON 
```

gdt commented:
@pnorman @alex85k  I've rerun with alex85k's exported variables, and then looked at ktrace to see what NAMI lookups were made.  It seems that the gcc 4.8 include files are being found, but then it dawned on me that I have boost built with the system gcc (4.5).   And since you can't mix gcc versions with C++, that's not going to work.   So this means I'd have to rebuild boost with 48.   Since this did work fine on NetBSD 7 which has gcc 48 native (and thus boost/etc.) were built with it, I'm going to declare victory.

I don't know if it's my lack of familiarity with cmake, or the docs, but it might be good for the README.md to explain what has to be put in which cmake variables to support alternative compilers and more importantly dependencies in a prefix other than /usr.

I also wonder if all of the exported variables alex85k put in are really necessary.  While i can see the point, it seems like it shouldn't be required (on redhat or on my netbsd system).

gdt commented:
Replying to my own complaint about too many variables, after reading followups (sorry, should have before), I trimmed down to this:

GCC_HOME=/usr/pkg/gcc48
export CC=${GCC_HOME}/bin/gcc CXX=${GCC_HOME}/bin/g++
export CMAKE_PREFIX_PATH=/usr/pkg
cmake -DBUILD_TESTS=ON 

and things are ok, modulo format string warnings (surely not a cmake issue) and the std:to_string complaint (which I attribute to boost compiled with gcc 4.5).   Actually the following is ok too - it finds gcc 4.8 and also postgresql (from /usr/pkg):

PATH=/usr/pkg/gcc48/bin:$PATH
cmake .. -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTS=ON

So this looks fine to me in the end; thanks for listening and for the advice.

alex85k commented:
@gdt: Thank you for sharing your experience! We'll add the most common steps to success into Readme.md. For now there is only CMAKE_PREFIX_PATH and BOOST_ROOT there, we need to write something about changing compilers too.
According LD_LIBRARY_PATH, C_INLCUDE_PATH etc. - I have just copied them long ago from stackouverflow. As far as I remember, they help to build Autotools projects (pass its checks and so on) and cmake seems to be smart enough.

Or maybe you just still use headers and stdlibc++ from system - could you please check the pathes found in CMakeLists.txt ?

pnorman commented:
> @pnorman @alex85k I've rerun with alex85k's exported variables, and then looked at ktrace to see what NAMI lookups were made. It seems that the gcc 4.8 include files are being found, but then it dawned on me that I have boost built with the system gcc (4.5). And since you can't mix gcc versions with C++, that's not going to work. So this means I'd have to rebuild boost with 48. Since this did work fine on NetBSD 7 which has gcc 48 native (and thus boost/etc.) were built with it, I'm going to declare victory.

Yes, if you want to run modern software on an OS that doesn't have C++11 and an appropriate STL you often have to build **everything** yourself. I've done it on Centos 6 before devtoolset-2, building gcc, autotools, boost, etc, etc.

> I don't know if it's my lack of familiarity with cmake, or the docs, but it might be good for the README.md to explain what has to be put in which cmake variables to support alternative compilers and more importantly dependencies in a prefix other than /usr.

We should refer to the appropriate cmake help page, not try to explain all cmake options and options only needed with older OSes and custom source installls.

pnorman commented on README.md:
````diff
@@ -48,42 +48,83 @@ Both GCC 4.8 and Clang 3.4 meet this requirement.
+sudo yum install geos-devel proj-devel proj-epsg
+```
+
+If the default version of PostgreSQL is too old, the official RPM repository
````
This is getting into how to install PostgreSQL, not building osm2pgsql. What version does Centos 7 have?

alex85k replied:
It has postgresql 9.2 that should be enough to build client, but Proj and Geos are only available via epel. For testing user also had to install posgis from somewhere.
You think that mentioning yum.postgresql.org would be enough?

pnorman replied:
If we can't condense it down I'd leave it as something like `For most uses, PostGIS will need to be installed but is not packaged on Redhat/CentOS`, adding a link as appropriate.

pnorman commented:
The building section of the readme has about doubled in length, going from 57 to 98 lines. `README.md` is an overview. If we need in-depth instructions we can have them separate in their own file, but it looks like most of the options should be able to be explained by giving an appropriate cmake help command or pointing to a cmake page.

alex85k commented:
I did not find CMAKE_PREFIX_PATH for over a year with all that "logical" help pages :)
Maybe we should create wiki page for that building & testing advices? Then we can remove "fresh postgresql" and "custom libraries" sections from readme without danger of having tens tickets like https://github.com/Project-OSRM/osrm-backend/issues?q=is%3Aissue+cmake+is%3Aclosed  :)

gdt commented:
alex85k <notifications@github.com> writes:

> @gdt: Thank you for sharing your experience! We'll add the most common
> steps to success into Readme.md. For now there is only
> CMAKE_PREFIX_PATH and BOOST_ROOT there, we need to write something
> about changing compilers too.  According LD_LIBRARY_PATH,
> C_INLCUDE_PATH etc. - I have just copied them long ago from
> stackouverflow. As far as I remember, they help to build Autotools
> projects (pass its checks and so on) and cmake seems to be smart
> enough.
>
> Or maybe you just still use headers and stdlibc++ from system - could
> you please check the pathes found in CMakeLists.txt ?

I wasn't able to figure out paths, but libstdc++ from the system is from
gcc 4.5, so that surely isn't going to work.  Paul is quite right that
for code that needs C++11, one has to install gcc 4.8 or higher and
build everything with it.  That's unfortunate; due to OS release cycles
and conservatism about moving to newer compilers, systems that can't
fairly be called crufty have < 4.8, even though 4.8.1 has been out 2.5
years.  All that said, it's clear that the issues are about any program
in C++11, and not about osm2pgsql in particular.  Hence I will withdraw
my trouble report, because it's clear what I should do to build on a
system with gcc 4.5.

Thanks for the help; I quickly came to understand the real issues from
all the helpful comments.

pnorman commented on .gitignore:
```diff
@@ -1,34 +1,7 @@
+build
 
 log
 test-driver
```
`test-driver` is from autotools, does CMake use the same name?

DerDakon replied:
When you run test through CTest the logs and stuff will be stored in Testing/ inside the build directory. Since in-source builds are forbidden and "build" is already ignored I don't think anything special has to be added.

pnorman commented on .gitignore:
```diff
@@ -58,9 +31,5 @@ tests/*.log
 tests/*.trs
```
are `*.trs` files still use with CMake?

pnorman commented:
I cleaned up some of the readme with alex85k#1

pnorman commented:
With a standard build (``cmake ..``, no other options), how do I

* Build the tests, but not run them
* Run all the tests
* Run *a* test

pnorman commented on .travis.yml:
```diff
@@ -44,17 +43,17 @@ install:
-  - ./configure
-  - make -j2 check TESTS='' CXXFLAGS="$BUILDCXXFLAGS"
+  - mkdir build && cd build
+  - cmake .. -DBUILD_TESTS=ON -DCMAKE_BUILD_TYPE=Debug $CMAKEOPTS
```
Is `$CMAKEOPTS` set anywhere? I can't find it elsewhere in the diff

alex85k replied:
This can be needed if we test different options on Travis, but currently it can be removed.

pnorman replied:
I think it's clearer if we remove it

alex85k commented:
>With a standard build (cmake .., no other options), how do I

>Build the tests, but not run them
>Run all the tests
>Run a test

No tests targets are generated until ``cmake .. -DBUILD_TESTS=ON``. If we want them to build under special target or run, EXCLUDE_FROM_ALL can be used.
http://stackoverflow.com/questions/3290280/cmake-and-ctest-how-to-make-target-nightly-actually-build-the-tests

Running all tests:
```
make test
(calls ctest)
```

Running custom test:
```
make test ARGS="-VV -L NoDB"     # run by label
make test ARGS="-VV -R hstore"    # run by substring of test name

```

pnorman commented:
> No tests targets are generated until cmake .. -DBUILD_TESTS=ON. If we want them to be available for custom targets, EXCLUDE_FROM_ALL option can be used.
> http://stackoverflow.com/questions/3290280/cmake-and-ctest-how-to-make-target-nightly-actually-build-the-tests (but I could not remember good example of using it)

I'd prefer if we always generated the test target, but left them out of make all.

Aside from that, I'm happy with it though I might have a go at improving the docs for running the tests.

alex85k commented:
Here is the "make check". If ``BUILD_TESTS=ON``, all tests are built by ``make`` .
If ``BUILD_TESTS=OFF``,  then ``make`` builds only main executable, while ``make check`` builds and runs tests.

Passing parameters to custom targets seem to be impossible. Forcing ``make test`` build the tests is possible but tricky. @DerDakon ? :)

DerDakon commented:
If I get it right your idea is "mkdir build && cd build && cmake .. && make test" and everything is built and tests are run. I don't know if that is possible, at least not in a portable way. I don't think it's worth the hassle anyway. If you want to run the tests build the software first.

pnorman commented:
> If I get it right your idea is "mkdir build && cd build && cmake .. && make test" and everything is built and tests are run.

The key is that choosing to build the tests or not and to run the tests or not is a **build-time** choice, not a configuration time (cmake) choice. 

It sounds like this is the right behavior with `make check` and `BUILD_TESTS=OFF`, but I'm testing now

pnorman commented:
> Here is the "make check". If ``BUILD_TESTS=ON``, all tests are built by ``make`` .
If ``BUILD_TESTS=OFF``,  then ``make`` builds only main executable, while ``make check`` builds and runs tests.``make test`` works as usual and can not be modified.

No, with `cmake .. -DBUILD_TESTS=OFF` when I try `make check` I get "*** No rule to make target 'check'.  Stop."

pnorman commented:
> No, with cmake .. -DBUILD_TESTS=OFF when I try make check I get "*** No rule to make target 'check'. Stop."

I might not have done a `git fetch`, will double-check

gdt commented:
Paul Norman <notifications@github.com> writes:

>> If I get it right your idea is "mkdir build && cd build && cmake
>> .. && make test" and everything is built and tests are run.
>
> The key is that choosing to build the tests or not and to run the
> tests or not is a **build-time** choice, not a configuration time
> (cmake) choice.

I think this is a bug.  People should always be encouraged to run the
tests.   So it would be nice if the build defaulted to building tests
unless they are configured off.

DerDakon commented:
Just change the default of the BUILD_TESTS option to "On" then.

alex85k commented:
@pnorman: Does ``make check`` work now. I am enabling BUILD_TESTS by default as there are no extra dependencies and tests are rather fast to build. Let me know if it should be reverted :)

pnorman commented:
> By the way, it seems that Appveyor is not turned on here...

It is, but it wasn't picking up this PR. Perhaps because this PR predates Appveyor.

I'm moving around some Appveyor names, so that might change things too.

pnorman commented:
I'd prefer `BUILD_TESTS` off by default, but asking @lonvia.

With `BUILD_TESTS=OFF` I'm having trouble telling it to build the tests without telling it to also run them.

I think this is ready, though I'm sure there will be follow-up PRs

DerDakon commented:
When you have configured (and maybe even built) the package and want to enable the tests:
```Shell
cd build
cmake -D BUILD_TESTS=On .
make
```

pnorman commented:
merged 17dfc1ec3f7ada36004dde403281c3582fceb5b5

alex85k commented:
With special thanks for @DerDakon for great CMake experience!

alex85k commented:
What should we do with tests?
https://github.com/alex85k/osm2pgsql/commit/aa9e7999fb421f59f4b7b9ad7dde961b33e1359f
https://github.com/alex85k/osm2pgsql/commit/e7a9a6ef17c4b62389094cc201194987a949c7f2

At least these parts:
https://github.com/alex85k/osm2pgsql/commit/aa9e7999fb421f59f4b7b9ad7dde961b33e1359f
https://github.com/alex85k/osm2pgsql/commit/e7a9a6ef17c4b62389094cc201194987a949c7f2
needs including.

pnorman commented:
Fixing in #479
